projects/graph/src/draw.py

```diff
@@ -1,14 +1,240 @@
+    self.plot = figure(title=title, x_range=(0, width), y_range=(0, height))
+    self.plot.axis.visible = show_axis
+    self.plot.grid.visible = show_grid
+    self._setup_graph_renderer(circle_size, draw_components)
+    self._setup_labels()
+
+    def _setup_graph_renderer(self, circle_size, draw_components):
+        # The renderer will have the actual logic for drawing
+        graph_renderer = GraphRenderer()
+        # Saving vertices in an arbitrary but persistent order
+        self.vertex_list = list(self.graph.vertices.keys())
+        # Add the vertex data as instructions for drawing nodes
+        graph_renderer.node_renderer.data_source.add(
+            [vertex.label for vertex in self.vertex_list], "index"
+        )
+        colors = (
+            self._get_connected_component_colors()
+            if draw_components
+            else self._get_random_colors()
+        )
+        graph_renderer.node_renderer.data_source.add(colors, "color")
+        # And circles
+        graph_renderer.node_renderer.glyph = Circle(
+            size=circle_size, fill_color="color"
+        )
+        # Add the edge [start, end] indices as instructions for drawing edges
+        graph_renderer.edge_renderer.data_source.data = self._get_edge_indexes()
+        self.randomize()  # Randomize vertex coordinates, and set as layout
+        graph_renderer.layout_provider = StaticLayoutProvider(graph_layout=self.pos)
+        # Attach the prepared renderer to the plot so it can be shown
+        self.plot.renderers.append(graph_renderer)
+
+    def _get_random_colors(self, num_colors=None):
+        colors = []
+        num_colors = num_colors or len(self.graph.vertices)
+
+        for _ in range(num_colors):
+            color = "#" + "".join([choice("0123456789ABCDEF") for j in range(6)])
+            colors.append(color)
+        return colors
+
+        def _get_edge_indexes(self):
+            start_indices = []
+            end_indices = []
+            checked = set()
+
+            for vertex, edges in self.graph.vertices.items():
+                if vertex not in checked:
+                    for destination in edges:
+                        start_indices.append(vertex.label)
+                        end_indices.append(destination.label)
+                    checked.add(vertex)
+                return dict(start=start_indices, end=end_indices)
+
+    def _setup_labels(self):
+        label_data = {"x": [], "y": [], "names": []}
+
+        for vertex_label, (x_pos, y_pos) in self.pos.items():
+            label_data["x"].append(x_pos)
+            label_data["y"].append(y_pos)
+            label_data["names"].append(vertex_label)
+        label_source = ColumnDataSource(label_data)
+        labels = LabelSet(
+            x="x",
+            y="y",
+            text="names",
+            level="glyph",
+            text_align="center",
+            text_baseline="middle",
+            source=label_source,
```

Comment:
Fantastic formatting! 🥇
It's clear, conconcise and well labeled with comments. This is the kind of extra step that will help you get far in coding. You don't just churn out junk, you make sure it's presentable. Way to go!